### ISSUE TITLE
FR: [aarch64 linux]

### ISSUE THREAD

**Issue by marek-lach** [koreader/koreader#6782]
The Debian builds for KOReader could hopefully include the arm64 architecture too, so that it can be used on mobile devices with linux as well, like the PinePhone, or the Librem5.

**Comment by pazos** [comment]
IIRC there's something on crosscompile that hardcodes `-m32`. But a native compilation using `./kodev release debian` on a aarch64 should build and package fine.

Anyways, Do you have a arm64 debian device to test @marek-lach ?

**Comment by NiLuJe** [comment]
FWIW, everything should be ARM64 ready, we took care of that in another issue about the same for Android that I can't quite find right now ;p.

**Comment by marek-lach** [comment]
> IIRC there's something on crosscompile that hardcodes `-m32`. But a native compilation using `./kodev release debian` on a aarch64 should build and package fine.
> 
> Anyways, Do you have a arm64 debian device to test @marek-lach ?

Yes, I do have the PinePhone with Debian. I could test it, and give feedback if it's made ready by someone else, but am not a programmer myself :-)

**Comment by pazos** [comment]
> Yes, I do have the PinePhone with Debian. I could test it, and give feedback if it's made ready by someone else, but am not a programmer, nor a package maintainer myself :-)

That's cool. I will ping you if I manage to build one.

> FWIW, everything should be ARM64 ready, we took care of that in another issue about the same for Android that I can't quite find right now ;p.

@NiLuJe thanks for the reminder. It is in https://github.com/koreader/koreader-base/issues/993#issuecomment-544212641

**Comment by pazos** [comment]
I got a build without much trouble. Only two things to tweak (I'll leave that to @NiLuJe). Here I include the changes I did to build a test deb package for aarch64.


### Leptonica: disable fmemopen symbol patch on aarch64.

```diff
diff --git a/thirdparty/leptonica/CMakeLists.txt b/thirdparty/leptonica/CMakeLists.txt
index 8cdf649..7422f1a 100644
--- a/thirdparty/leptonica/CMakeLists.txt
+++ b/thirdparty/leptonica/CMakeLists.txt
@@ -26,9 +26,7 @@ ko_write_gitclone_script(
 # NOTE: The patch will be much simpler w/ the upcoming leptonica version,
 # there's a 1.74 version of the patch already in our tree for when that time
 # comes ;).
-if(${ARM_GLIBC_GTE_2_22})
-    set(PATCH_CMD "${KO_PATCH} ${CMAKE_CURRENT_SOURCE_DIR}/leptonica-1.74.1-fmemopen-arm-compat-symbol.patch")
-endif()
```
### Change openssl platform

```diff
diff --git a/thirdparty/openssl/CMakeLists.txt b/thirdparty/openssl/CMakeLists.txt
index 58b1574..ec768ac 100644
--- a/thirdparty/openssl/CMakeLists.txt
+++ b/thirdparty/openssl/CMakeLists.txt
@@ -45,11 +45,7 @@ if($ENV{ANDROID})
         set(CFG_OPTS "android-x86 ${CFG_OPTS}")
     endif()
 else()
-    # If we're on ARM, make it so
-    if (${CHOST} MATCHES "^arm-.*")
-        # Yes, OpenSSL's preset names make no sense.
-        set(CFG_OPTS "linux-armv4 -DL_ENDIAN ${CFG_OPTS}")
-    endif()
+    set(CFG_OPTS "linux-aarch64 -DL_ENDIAN ${CFG_OPTS}")
 endif()
 
 # Do we support SIMD?
```

**Comment by pazos** [comment]
And now the proper diff for the new target:

### frontend

```diff
diff --git a/Makefile b/Makefile
index 6351cf00..db6be5db 100644
--- a/Makefile
+++ b/Makefile
@@ -533,6 +533,9 @@ else ifeq ($(TARGET), debian-armel)
 else ifeq ($(TARGET), debian-armhf)
        make debianupdate
        $(CURDIR)/platform/debian/do_debian_package.sh $(INSTALL_DIR) armhf
+else ifeq ($(TARGET), debian-arm64)
+       make debianupdate
+       $(CURDIR)/platform/debian/do_debian_package.sh $(INSTALL_DIR) arm64
 else ifeq ($(TARGET), macos)
        make macosupdate
        $(CURDIR)/platform/mac/do_mac_bundle.sh $(INSTALL_DIR)
```
```diff
diff --git a/kodev b/kodev
index b989678e..bd5cc44c 100755
--- a/kodev
+++ b/kodev
@@ -132,6 +132,7 @@ SUPPORTED_TARGETS="
     debian          Debian package for current arch
     debian-armel    Debian package for generic armel devices
     debian-armhf    Debian package for generic armhf devices
+    debian-arm64    Debian package for generic 64 bits arm devices (aarch64)
     macos           MacOS app bundle. You need a mac to build this package
     emu             (*default) If no TARGET is given, assume emulator
     win32
@@ -251,6 +252,10 @@ ${SUPPORTED_TARGETS}"
             make TARGET=debian-armhf
             assert_ret_zero $?
             ;;
+        debian-arm64)
+            make TARGET=debian-arm64
+            assert_ret_zero $?
+            ;;
         macos)
             is_mac
             make TARGET=macos
@@ -348,6 +353,9 @@ ${SUPPORTED_TARGETS}"
         debian-armhf)
             make TARGET=debian-armhf clean
             ;;
+        debian-arm64)
+            make TARGET=debian-arm64 clean
+            ;;
         macos)
             is_mac
             make TARGET=macos clean
@@ -485,6 +493,10 @@ ${SUPPORTED_RELEASE_TARGETS}"
             kodev-build debian-armhf
             make TARGET=debian-armhf update
             ;;
+        debian-arm64)
+            kodev-build debian-arm64
+            make TARGET=debian-arm64 update
+            ;;
         macos)
             is_mac
             kodev-build macos
```

### base

```diff
diff --git a/Makefile.defs b/Makefile.defs
index 7f99ad2..49f1458 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -122,6 +122,10 @@ else ifeq ($(TARGET), debian-armhf)
        CHOST?=arm-linux-gnueabihf
        export SDL=1
        export DEBIAN_CROSS=1
+else ifeq ($(TARGET), debian-arm64)
+       CHOST?=aarch64-linux-gnu
+       export SDL=1
+       export DEBIAN_CROSS=1
 else ifeq ($(TARGET), macos)
        export SDL=1
        export MACOS=1
@@ -337,6 +341,8 @@ ARMV6_1136_ARCH:=-march=armv6j -mtune=arm1136jf-s -mfpu=vfp -marm
 ARMV6_GENERIC_ARCH:=-march=armv6 -mtune=generic-armv6 -marm
 # Generic armv7
 ARMV7_GENERIC_ARCH:=-march=armv7-a -mtune=generic-armv7-a -mfpu=vfpv3 -mthumb
+# Generic arm 64 bits
+ARM64_GENERIC_ARCH:=-march=armv8-a
 # Cortex A8 (K4, Kindle Touch, PW1, Kobos since the Touch)
 ARMV7_A8_ARCH:=-march=armv7-a -mtune=cortex-a8 -mfpu=neon -mthumb
 # Cortex A9 (Kindle PW2)
@@ -437,6 +443,8 @@ else ifeq ($(TARGET), debian-armel)
 else ifeq ($(TARGET), debian-armhf)
        ARM_ARCH:=$(ARMV7_GENERIC_ARCH)
        ARM_ARCH+=-mfloat-abi=hard
+else ifeq ($(TARGET), debian-arm64)
+       ARM_ARCH:=$(ARM64_GENERIC_ARCH)
 else
        # else, we are compiling emulator on x86
        ARM_ARCH:=
```

@marek-lach: here's a test build: https://www.dropbox.com/s/fus72zbgtxeal36/koreader-2020.03.2-401-arm64.deb?dl=0

Please install and report if it works.

**Comment by marek-lach** [comment]
@pazos It does work, but only after fixing some dependency issues with the _-f_ parameter after installation from the _.deb_ file. This is what it looks like:

Get:1 http://deb.debian.org/debian bullseye/main arm64 fonts-noto-hinted all 20200323-1
Get:2 http://deb.debian.org/debian bullseye/main arm64 fonts-droid-fallback all 1:6.0.1r16-1.1

These dependencies should be listed as required, after that the application works fine.

**Comment by pazos** [comment]
> These dependencies should be listed as required, after that the application works fine.

They [are](https://github.com/koreader/koreader/blob/master/platform/debian/do_debian_package.sh#L53), as that's why apt / dpkg can fix what's missing.

**Comment by pazos** [comment]
I guess we can replace armel monthly builds with arm64 ones. Probably nobody uses KOReader on soft float EABI anyways.

I think even a raspbian compatible armhf makes more sense than armel. Even more: Debian armel port targets really old armv4 boards, while our armel port only covers armv6+

**Comment by marek-lach** [comment]
Quick question: why not submit koreader also into the official Debian repositories (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=969266), since you already have the packages made for the various architectures? It would certainly make koreader easier to install/update on Debian based systems...

**Comment by poire-z** [comment]
Because it would be a lot bigger responsabilities and constraints? :) 
And we would probably be required to not ship "every library" and make use of those from other packages, which would be a lot tougher and time consuming in keeping all that working and updated , and a different build process? (https://github.com/koreader/koreader/issues/3777#issuecomment-374399064).

**Comment by NiLuJe** [comment]
What he said ;). Hard pass, unless a Debian fanatic volunteers to actually do the work, and it doesn't make things a horrible mess for the 99.998% rest of the world.

**Comment by pazos** [comment]
> Quick question: why not submit koreader also into the official Debian repositories (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=969266), since you already have the packages made for the various architectures? It would certainly make koreader easier to install/update on Debian based systems...

There's no benefit in being supported by Debian. Just a dll/so hell.
There are a few things that will be worst (aside from the dynamic library hell): the size of the package will be reduced from 30MB to 14MB (hello translations!), but the size of dependencies will grow from a few MBs to +250MB (hello noto-hinted-cjk-extra!)

The thing we can do to make the install/update easier is holding a litle repo in apt.koreader.rocks that can be added to sources list, so install/update is a matter of `apt update && apt install koreader` / `apt update && apt uprgrade`

The amount of work needed to devendor libraries in koreader-base is **huge**. The benefit is 0. The "solution" offered to https://github.com/koreader/koreader/issues/3108 was to package the emulator, without debug symbols, in a debian friendly manner. Was almost 0 effort and does work reliabily enough.

**Comment by marek-lach** [comment]
> There's no benefit in being supported by Debian. Just a dll/so hell.
> There are a few things that will be worst (aside from the dynamic library hell): the size of the package will be reduced from 30MB to 14MB (hello translations!), but the size of dependencies will grow from a few MBs to +250MB (hello noto-hinted-cjk-extra!)
> 
> The thing we can do to make the install/update easier is holding a litle repo in apt.koreader.rocks that can be added to sources list, so install/update is a matter of `apt update && apt install koreader` / `apt update && apt uprgrade`
> 
> The amount of work needed to devendor libraries in koreader-base is **huge**. The benefit is 0. The "solution" offered to #3108 was to package the emulator, without debug symbols, in a debian friendly manner. Was almost 0 effort and does work reliabily enough.

A very nice, and detailed explanation, that makes sense. Thanks for your time writing this!

**Comment by marek-lach** [comment]
@Frenzie Is this solved by either the armel.deb, or the armhf.deb release, those are different architectures, no? I don't actually see an `arm64.deb` release, although the test build @pazos did worked just fine with my testing, and could be released as-is.

**Comment by pazos** [comment]
This was solved by adding support to build the package from source.
Now to build a package in the build server https://github.com/koreader/virdevenv/pull/56 is needed, then @Frenzie or somebody else (nah, it is always Frenzie :p, thanks!) need to do pull these changes, build a new docker container, push the container somewhere and modify the build scripts to generate a package.

Is still a proper target and you can build it from an Intel/AMD computer with `./kodev release debian-arm64` or even from your target device with `./kodev release debian`  as some aarch64 devices are quite capable.

**Comment by Frenzie** [comment]
@pazos I thought you weren't fully done with it yet. I was waiting for you to merge it when ready. ^_^

**Comment by Frenzie** [comment]
Unless something goes wrong, you'll be able to grab the artifact from https://gitlab.com/koreader/nightly-builds/-/jobs/954027763 in a half hour or so. But I'll be off to bed by then.

**Comment by pazos** [comment]
@Frenzie: TBH I forgot about that PR :P

**Comment by Frenzie** [comment]
Anyway, I can't really test it (maybe in Qemu or something), but it looks like all went well. Does it run properly?

**Comment by pazos** [comment]
I can't test either. Pinging @marek-lach

**Comment by Eloitor** [comment]
I have a pinephone and I would like to use koreader in it. This issue is closed, but I don't find the build for aarch64 debian in the releases page... What can I do?

**Comment by Frenzie** [comment]
Are you saying the one called arm64 doesn't work for you?

**Comment by Eloitor** [comment]
It's working, thanks! I didn't know they are the compatible.
